9661: Mark post as sticky

danielebogo opened the pull request:
Fixes #9661 
This is PR introduces the **Mark a Post as sticky** feature.

![sticky-post-ex](https://user-images.githubusercontent.com/912252/44859713-991c1500-ac6c-11e8-86ca-51b4ae5bf7fe.gif)

First *WordPressKit* pod has been updated to the version *1.3*. It includes the new property used by the *RemotePost* model and its decode/encode methods have been updated to handle it.

I created a new Data Model version (80) to add the property *isStickyPost* to the *Post* entity. The model has been updated too and the *PostService* has been updated to get the value from the *RemotePost* and viceversa.

*PostSettingsViewController.m* includes the new enum case *PostSettingsSectionStickyPost* that defines the new tableview section. The section cell is a *SwitchTableViewCell* class and it updates the post object with its block.

I updated *PostCardTableViewCell.m* (also modifying *PostCardImageCell.xib* and *PostCardTextCell.xib*) to display/hide the *Sticky Icon and Label*.

**To test:**

- Install the App on a previous version to test the new Data Model transition
- Select a post, then switch on/off the *Mark post as sticky*
- Create a post, then switch on/off the *Mark post as sticky*

jklausa commented on WordPress/Classes/WordPress.xcdatamodeld/.xccurrentversion:
```diff
@@ -3,6 +3,6 @@
 <dict>
 	<key>_XCCurrentVersionName</key>
-	<string>WordPress 79.xcdatamodel</string>
+	<string>WordPress 80.xcdatamodel</string>
```
Nobody told you about it before, so it's by no means your fault — but it's customary to add a entry in `MIGRATIONS.md` every time we create a new Core Data model version. Would you be so kind and and one for this :)?

danielebogo replied:
Oh yes, now you said it I remember i did it once during the trial! Done! 😀

jklausa commented on Podfile:
```diff
@@ -23,7 +23,7 @@ def shared_with_networking_pods
     pod 'Alamofire', '4.7.2'
     pod 'wpxmlrpc', '0.8.3'
-    pod 'WordPressKit', '1.2.1'
+    pod 'WordPressKit', '1.3'
```
Argh, sorry for this :) I pushed out my own PR today that updates it to `'1.4.0-beta.1` — this is what's causing the merge conflict. Make sure it still points to `'1.4.0-beta.1` after you resolve the conflict :)

jklausa commented on WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m:
```diff
@@ -398,6 +401,7 @@ - (void)configureSections
                       @(PostSettingsSectionMeta),
                       @(PostSettingsSectionFormat),
                       @(PostSettingsSectionFeaturedImage),
+                      @(PostSettingsSectionStickyPost),
```
I think we need to take care whether the site is self-hosted or not here. I seem to remember the WPKit part only took care of Jetpack/WPCom sites, so for self-hosted sites we'd show an option here that users can't actually change.

danielebogo replied:
@jklausa all *Self-hosted* blogs or only *Self-hosted non-Jetpack* blogs?
Let me know if this is correct:
In the first case I can check if the `post.blog` is `isHostedAtWPcom`.
In the second one I can check `post.blog.capabilities` exists. *Self-hosted non-Jetpack* blogs have no capabilities. This check is used by the `AztecPostViewController` when a new Editor Context is created for a given Post.

jklausa replied:
Right, I think non-Jetpack blogs in this case!

danielebogo replied:
@jklausa done! 😄 Let me know it it's ok!

aerych commented on WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m:
```diff
@@ -396,16 +396,24 @@ - (BOOL)textFieldShouldReturn:(UITextField *)textField
+    // Self-hosted non-Jetpack blogs have no capabilities,
+    // so we'll remove the Sticky Post section.
+    //
+    if (self.post.blog.capabilities == nil) {
```
Since `capabilities` is something we get from the REST API, would it be better to check `![self.post.blog supportsRestApi]`?

danielebogo replied:
Yep! It makes sense! I used the same logic I found into the AztecPostViewController when a context is created.

aerych commented:
Thanks @jklausa for thinking of self-hosted blogs. 
@danielebogo I just retested and everything looks good to me. :)

SergioEstevao commented:
It [looks that](https://codex.wordpress.org/XML-RPC_WordPress_API/Posts#wp.getPost) the self-hosted sites/XML-RPC-API also supports sticky posts are we not planning to support it on the app?

It should be relatively straightforward to implement this on WordPressKit -> PostServiceRemoteXMLRPC.m, just need to read and write the parameter?

mtias commented:
Just passing by, but I think the "sticky post" pin metaphor is better suited for English's "post" word, and it fails a bit in languages that translate "post" differently.

danielebogo commented:
I updated WordPressKit to 1.4.0-beta.3 to allow self-hosted blogs to use the Mark as sticky feature.
I also removed the icon from the post card cell. I tested this final changes running the app and mark/unmark a post for a WP.com, self-hosted, self-hosted non-Jetpack blogs.

SylvesterWilmott commented:
@danielebogo Could you add a screenshot of the post list with the `sticky` label without the icon please

danielebogo commented:
@SylvesterWilmott 

![simulator-screen-shot---iphone-8---2018-09-06-at-11 02 10](https://user-images.githubusercontent.com/912252/45150761-86519500-b1c4-11e8-87f1-340fe5fe6982.jpg)

danielebogo commented:
Hound is stuck for hours. I'll proceed to merge this PR.
